File: javascript/leetcode/average-of-levels-in-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var averageOfLevels = function(root) {
    const levelData = [];
    const traverse = function(root, levelIndex) {
        let data = levelData[levelIndex] || [];
        data.push(root.val);
        if (root.left) {
            traverse(root.left, levelIndex + 1);
        }

        if (root.right) {
            traverse(root.right, levelIndex + 1);
        }
        levelData[levelIndex] = data;
    };
    traverse(root, 0);
    return levelData.map((data) => data.reduce((n, acc) => n + acc, 0) / data.length);
};

File: java/leetcode/average-of-levels-in-binary-tree.java
import java.util.stream.Collectors;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Double>> levelData;
    public List<Double> averageOfLevels(TreeNode root) {
        levelData = new ArrayList<List<Double>>();
        traverse(root, 0);
        return levelData.stream().map(data -> data.stream().reduce(0.0, (n, acc) -> n + acc) / data.size()).collect(Collectors.toList());
    }

    public void traverse(TreeNode root, int levelIndex) {
        if (levelData.size() < levelIndex + 1) {
            levelData.add(new ArrayList<Double>());
        }
        List<Double> data = levelData.get(levelIndex);
        data.add(new Double(root.val));
        if (root.left != null) {
            traverse(root.left, levelIndex + 1);
        }

        if (root.right != null) {
            traverse(root.right, levelIndex + 1);
        }
        levelData.set(levelIndex, data);
    }
}
